lightning/src/ln/channel.rs

```diff
@@ -1977,28 +1992,52 @@ impl<Signer: WriteableEcdsaChannelSigner> Channel<Signer> {
 	}
 
 	pub fn get_update_fulfill_htlc_and_commit<L: Deref>(&mut self, htlc_id: u64, payment_preimage: PaymentPreimage, logger: &L) -> UpdateFulfillCommitFetch where L::Target: Logger {
+		let release_cs_monitor = self.pending_monitor_updates.iter().all(|upd| !upd.blocked);
 		match self.get_update_fulfill_htlc(htlc_id, payment_preimage, logger) {
-			UpdateFulfillFetch::NewClaim { mut monitor_update, htlc_value_msat, msg: Some(_) } => {
-				let mut additional_update = self.build_commitment_no_status_check(logger);
-				// build_commitment_no_status_check may bump latest_monitor_id but we want them to be
-				// strictly increasing by one, so decrement it here.
-				self.latest_monitor_update_id = monitor_update.update_id;
-				monitor_update.updates.append(&mut additional_update.updates);
-				self.monitor_updating_paused(false, true, false, Vec::new(), Vec::new(), Vec::new());
-				self.pending_monitor_updates.push(monitor_update);
+			UpdateFulfillFetch::NewClaim { mut monitor_update, htlc_value_msat, msg } => {
```

Comment:
This would be a much smaller diff to `main`:
```
	pub fn get_update_fulfill_htlc_and_commit<L: Deref>(&mut self, htlc_id: u64, payment_preimage: PaymentPreimage, logger: &L) -> UpdateFulfillCommitFetch where L::Target: Logger {
		let release_cs_monitor = self.pending_monitor_updates.iter().all(|upd| !upd.blocked);
		macro_rules! insert_update {
			($monitor_update: expr, $msg_is_some: expr) => {{
				// Even if we aren't supposed to let new monitor updates with commitment state
				// updates run, we still need to push the preimage ChannelMonitorUpdateStep no
				// matter what. Sadly, to push a new monitor update which flies before others
				// already queued, we have to insert it into the pending queue and update the
				// update_ids of all the following monitors.
				let insert_pos = self.pending_monitor_updates.iter().position(|upd| upd.blocked)
					.unwrap_or(self.pending_monitor_updates.len());
				let new_mon_id = self.pending_monitor_updates.get(insert_pos)
					.map(|upd| upd.update.update_id).unwrap_or($monitor_update.update_id);
				$monitor_update.update_id = new_mon_id;
				self.pending_monitor_updates.insert(insert_pos, PendingChannelMonitorUpdate {
					update: $monitor_update, blocked: false,
				});
				for held_update in self.pending_monitor_updates.iter_mut().skip(insert_pos + 1) {
					held_update.update.update_id += 1;
				}
				if !release_cs_monitor && $msg_is_some {
					debug_assert!(false, "If there is a pending blocked monitor we should have MonitorUpdateInProgress set");
					let update = self.build_commitment_no_status_check(logger);
					self.pending_monitor_updates.push(PendingChannelMonitorUpdate {
						update, blocked: true,
					});
				}
				insert_pos
			}}
		}
		match self.get_update_fulfill_htlc(htlc_id, payment_preimage, logger) {
			UpdateFulfillFetch::NewClaim { mut monitor_update, htlc_value_msat, msg: Some(_) } => {
				let mut additional_update = self.build_commitment_no_status_check(logger);
				// build_commitment_no_status_check may bump latest_monitor_id but we want them to be
				// strictly increasing by one, so decrement it here.
				self.latest_monitor_update_id = monitor_update.update_id;
				monitor_update.updates.append(&mut additional_update.updates);
				self.monitor_updating_paused(false, true, false, Vec::new(), Vec::new(), Vec::new());
				let insert_idx = insert_update!(monitor_update, true);
				UpdateFulfillCommitFetch::NewClaim {
					monitor_update: &self.pending_monitor_updates.get(insert_idx).unwrap().update,
					htlc_value_msat,
				}
			},
			UpdateFulfillFetch::NewClaim { mut monitor_update, htlc_value_msat, msg: None } => {
				self.monitor_updating_paused(false, false, false, Vec::new(), Vec::new(), Vec::new());
				let insert_idx = insert_update!(monitor_update, false);
				UpdateFulfillCommitFetch::NewClaim {
					monitor_update: &self.pending_monitor_updates.get(insert_idx).unwrap().update,
					htlc_value_msat,
				}
			}
			UpdateFulfillFetch::DuplicateClaim {} => UpdateFulfillCommitFetch::DuplicateClaim {},
		}
	}
```
Seems like we can unconditionally do the weird insertion, since it'll just end up doing the current behavior if nothing's blocked?